src/dns.rs

```diff
@@ -0,0 +1,464 @@
+use crate::topology::config::GlobalOptions;
+use futures::{Async, Future, Poll};
+use hyper::client::connect::dns::{Name, Resolve};
+use std::io;
+use std::net::ToSocketAddrs;
+use std::net::*;
+use std::vec::IntoIter;
+use trust_dns_resolver::config::*;
+use trust_dns_resolver::error::ResolveErrorKind;
+use trust_dns_resolver::AsyncResolver;
+use trust_dns_resolver::BackgroundLookupIp;
+
+/// Default port for DNS service.
+const DNS_PORT: u16 = 53;
+
+/// Default version will behave identically to std::net::ToSocketAddrs implementations.
+#[derive(Clone, Default)]
+pub struct DnsResolver(Option<AsyncResolver>);
+
+impl DnsResolver {
+    /// (Self,errors)
+    pub fn new(config: &GlobalOptions) -> Result<Self, String> {
+        if config.dns.is_empty() {
+            return Ok(DnsResolver::default());
+        }
+
+        let mut resolve_config = ResolverConfig::new();
+
+        let mut errors = Vec::new();
+
+        for s in config.dns.iter() {
+            let parsed = s
+                .parse::<SocketAddr>()
+                .or_else(|_| s.parse::<IpAddr>().map(|ip| SocketAddr::new(ip, DNS_PORT)));
+            match parsed {
+                Ok(socket_addr) => resolve_config.add_name_server(NameServerConfig {
+                    socket_addr,
+                    protocol: Protocol::Udp,
+                    tls_dns_name: None,
+                }),
+                Err(error) => errors.push(format!(
+                    "DNS server IP {:?} is invalid. Parse Error: {:?}",
+                    s, error
+                )),
+            }
+        }
+
+        if !errors.is_empty() {
+            return Err(format!(
+                "DNS resolver initialization failed with errors: {:?}",
+                errors
+            ));
+        }
+
+        let mut options = ResolverOpts::default();
+        options.attempts = 2;
+        options.validate = false;
+        options.ip_strategy = LookupIpStrategy::Ipv4AndIpv6;
+        options.use_hosts_file = false;
+        options.distrust_nx_responses = false;
+        options.num_concurrent_reqs = 1;
+
+        let (resolver, worker) = AsyncResolver::new(resolve_config, options);
+
+        // spin up worker thread for worker future
+        std::thread::spawn(move || {
+            // Worker futer will finish when all DnsResolvers are dropped.
+            // Then this thread will also end.
+            tokio::runtime::current_thread::run(worker);
+        });
+
+        Ok(DnsResolver(Some(resolver)))
+    }
+
+    /// Resolves host:port address
+    /// Should be only called from outside tokio. From inside tokio, Async version should be used.
+    pub fn resolve_address(&self, s: &str) -> Result<IntoIter<SocketAddr>, io::Error> {
+        // Try to parse as a regular SocketAddr first
+        if let Some(addr) = s.parse().ok() {
+            return Ok(vec![addr].into_iter());
+        }
+
+        // Now 's' should contain name:port
+        let mut parts = s.rsplitn(2, ':');
```

Comment:
would a regex be better here?

Reply:
I think it would be a bit too much.

Reply:
is port always going to be required? if not, would an ipv6 addr mess with this?

(btw, given that everyone seems to want to include ports in the resolution, we should probably just add a method for this directly to the trust-dns-resolver)

Reply:
> is port always going to be required?

For this method yes, generally no. Only some places of usage use this method, and this method is the only one which expects port. More accurately those that have configuration strings that should contain host:port. 

And for including a method with port into trust-dns-resolver, it would be convenient, but in this case I would still stick to using method with only Name, otherwise new method stack, now with ex.(Name,port), would have to be implemented.